File: src/sort/mergesort.ts
// LombokAlgoritma — Bottom-Up Iterative Mergesort
// Apache-2.0 — @codinglombok
// O(n log n) all cases, O(n) space, Stable, No recursion stack

import type { CompareFn } from '../core/types.js';
import { defaultCompareFn } from '../core/types.js';

export function mergesort<T>(
  arr: T[],
  compareFn: CompareFn<T> = defaultCompareFn as CompareFn<T>,
  { inPlace = true }: { inPlace?: boolean } = {},
): T[] {
  const a = inPlace ? arr : [...arr];
  const n = a.length;
  if (n <= 1) return a;
  const tmp = new Array<T>(n);
  for (let width = 1; width < n; width *= 2) {
    for (let lo = 0; lo < n; lo += 2 * width) {
      const mid = Math.min(lo + width, n);
      const hi = Math.min(lo + 2 * width, n);
      // Merge a[lo..mid) and a[mid..hi) into tmp
      let i = lo, j = mid, k = lo;
      while (i < mid && j < hi) {
        if (compareFn(a[i] as T, a[j] as T) <= 0) tmp[k++] = a[i++] as T;
        else tmp[k++] = a[j++] as T;
      }
      while (i < mid) tmp[k++] = a[i++] as T;
      while (j < hi) tmp[k++] = a[j++] as T;
      for (let x = lo; x < hi; x++) a[x] = tmp[x] as T;
    }
  }
  return a;
}

File: ports/rust/src/sort/mergesort.rs
pub fn mergesort<T:Ord+Clone>(a:&mut[T]){
    let n=a.len();if n<=1{return;}
    let mut t=a.to_vec();let mut w=1;
    while w<n{
        let mut lo=0;while lo<n{
            let mid=(lo+w).min(n);let hi=(lo+2*w).min(n);
            let(mut i,mut j,mut k)=(lo,mid,lo);
            while i<mid&&j<hi{if a[i]<=a[j]{t[k]=a[i].clone();i+=1;}else{t[k]=a[j].clone();j+=1;}k+=1;}
            while i<mid{t[k]=a[i].clone();i+=1;k+=1;}
            while j<hi{t[k]=a[j].clone();j+=1;k+=1;}
            lo+=2*w;
        }
        a.clone_from_slice(&t);w*=2;
    }
}
